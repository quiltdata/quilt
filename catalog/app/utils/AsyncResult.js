import * as R from 'ramda'

import tagged from 'utils/tagged'

const AsyncResult = tagged(['Init', 'Pending', 'Ok', 'Err'])

AsyncResult.prop = (name, ...args) => AsyncResult.mapCase({ Ok: R.prop(name) }, ...args)

AsyncResult.props = R.curryN(2, (names, ...args) =>
  names.reduce(
    (acc, name) => ({
      ...acc,
      [name]: AsyncResult.prop(name, ...args),
    }),
    {},
  ),
)

export default AsyncResult
