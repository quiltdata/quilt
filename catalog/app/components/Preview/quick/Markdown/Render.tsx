import * as React from 'react'
import * as Lab from '@material-ui/lab'

import { useMarkdownRenderer } from 'components/Preview/loaders/Markdown'
import Markdown from 'components/Preview/renderers/Markdown'
import type * as Model from 'model'
import AsyncResult from 'utils/AsyncResult'

import Skeleton from './Skeleton'

interface RenderProps {
  value: string
  handle: Model.S3.S3ObjectLocation
}

export default function Render({ value, handle }: RenderProps) {
  const result = useMarkdownRenderer(AsyncResult.Ok(value), handle)
  // `result` is never `Pending` or `Init`, because we pass the `AsyncResult.Ok` value
  // but it can be `Err` if some post-processing fails
  return AsyncResult.case({
    _: (x: { value?: Error }) => (
      <Lab.Alert severity="error">{x.value?.message || 'Unexpected state'}</Lab.Alert>
    ),
    Ok: (rendered: string) => (
      <React.Suspense fallback={<Skeleton />}>
        <Markdown rendered={rendered} />
      </React.Suspense>
    ),
  })(result)
}
