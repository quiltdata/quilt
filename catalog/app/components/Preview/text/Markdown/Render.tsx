import * as React from 'react'
import * as Lab from '@material-ui/lab'

import { useMarkdownRenderer } from 'components/Preview/loaders/Markdown'
import Markdown from 'components/Preview/renderers/Markdown'
import type * as Model from 'model'
import AsyncResult from 'utils/AsyncResult'

import Skeleton from './Skeleton'

interface RenderProps {
  contents: string
  handle: Model.S3.S3ObjectLocation
}

export default function Render({ contents, handle }: RenderProps) {
  const result = useMarkdownRenderer(AsyncResult.Ok(contents), handle)
  return AsyncResult.case({
    _: () => null,
    Err: (error: Error) => <Lab.Alert severity="error">{error.message}</Lab.Alert>,
    Pending: () => <Skeleton />,
    Ok: (rendered: string) => (
      <React.Suspense fallback={<Skeleton />}>
        <Markdown rendered={rendered} />
      </React.Suspense>
    ),
  })(result)
}
