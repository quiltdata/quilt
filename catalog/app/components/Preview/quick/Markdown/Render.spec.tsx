import * as React from 'react'
import renderer from 'react-test-renderer'

import AsyncResult from 'utils/AsyncResult'

import Render from './Render'

jest.mock(
  'constants/config',
  jest.fn(() => {}),
)

const useMarkdownRenderer = jest.fn()
jest.mock('components/Preview/loaders/Markdown', () => ({
  ...jest.requireActual('components/Preview/loaders/Markdown'),
  useMarkdownRenderer: jest.fn(() => useMarkdownRenderer()),
}))

jest.mock(
  'components/Preview/renderers/Markdown',
  () =>
    ({ rendered }: { rendered: string }) => (
      // eslint-disable-next-line react/no-danger
      <b dangerouslySetInnerHTML={{ __html: rendered }}>Markdown</b>
    ),
)

jest.mock(
  '@material-ui/lab',
  jest.fn(() => ({
    Alert: ({ children }: { children: string }) => <div>Error: {children}</div>,
  })),
)

const handle = {
  bucket: 'foo',
  key: 'bar',
}

describe('app/components/Preview/quick/Render.spec.tsx', () => {
  it('it shows the error for Init state, because it is intended to run with already resolved value', () => {
    useMarkdownRenderer.mockReturnValue(AsyncResult.Init())
    const tree = renderer.create(<Render {...{ handle, value: 'any' }} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('it shows the error for Pending state, because it is intended to run with already resolved value', () => {
    useMarkdownRenderer.mockReturnValue(AsyncResult.Pending())
    const tree = renderer.create(<Render {...{ handle, value: 'any' }} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('returns error on Err', () => {
    useMarkdownRenderer.mockReturnValue(AsyncResult.Err(new Error('some error')))
    const tree = renderer.create(<Render {...{ handle, value: 'any' }} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('returns markdown on data', () => {
    useMarkdownRenderer.mockReturnValue(AsyncResult.Ok('<h1>It works</h1>'))
    const tree = renderer.create(<Render {...{ handle, value: 'any' }} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
