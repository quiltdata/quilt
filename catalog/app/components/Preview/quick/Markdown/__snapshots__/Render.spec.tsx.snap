// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/Preview/quick/Render.spec.tsx it shows the error for Init state, because it is intended to run with already resolved value 1`] = `
<div>
  Error: 
  Unexpected state
</div>
`;

exports[`app/components/Preview/quick/Render.spec.tsx it shows the error for Pending state, because it is intended to run with already resolved value 1`] = `
<div>
  Error: 
  Unexpected state
</div>
`;

exports[`app/components/Preview/quick/Render.spec.tsx returns error on Err 1`] = `
<div>
  Error: 
  some error
</div>
`;

exports[`app/components/Preview/quick/Render.spec.tsx returns markdown on data 1`] = `
<b
  dangerouslySetInnerHTML={
    {
      "__html": "<h1>It works</h1>",
    }
  }
>
  Markdown
</b>
`;
